File: basic-proxy-impl/src/main/java/com/study/proxy/impl/util/MethodFilter.java
package com.study.proxy.impl.util;

import org.objectweb.asm.Type;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.Arrays;
import java.util.List;
import java.util.function.BooleanSupplier;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class MethodFilter {

    public final Predicate<Method> notObjectMethod =
            (method) -> {
                String name = method.getName();
                String descriptor = Type.getMethodDescriptor(method);

                BooleanSupplier isEqualsMethod =
                        () -> "equals".equals(name) && "(Ljava/lang/Object;)Z".equals(descriptor);

                BooleanSupplier isHashCodeMethod =
                        () -> "hashCode".equals(name) && "()I".equals(descriptor);

                BooleanSupplier isToStringMethod =
                        () -> "toString".equals(name) && "()Ljava/lang/String;".equals(descriptor);
                return !(isEqualsMethod.getAsBoolean() ||
                        isToStringMethod.getAsBoolean() ||
                        isHashCodeMethod.getAsBoolean());
            };

    /**
     * Static methods and Object-methods will be ignored
     *
     * @param methods original methods
     * @return filtered methods
     */
    public List<Method> filter(Method[] methods) {
        return Arrays.stream(methods).
                filter(method -> !Modifier.isStatic(method.getModifiers())).
                filter(notObjectMethod).
                collect(Collectors.toList());
    }
}

File: basic-proxy-kotlin-impl/src/main/kotlin/com/study/proxy/impl/util/MethodFilter.kt
package com.study.proxy.impl.util

import org.objectweb.asm.Type
import java.lang.reflect.Method
import java.lang.reflect.Modifier

object MethodFilter {

    val notObjectMethod: (Method) -> Boolean = { method ->
        val name = method.name
        val descriptor = Type.getMethodDescriptor(method)

        val isEqualsMethod = { ("equals" == name && "(Ljava/lang/Object;)Z" == descriptor) }
        val isHashCodeMethod = { ("hashCode" == name && "()I" == descriptor) }
        val isToStringMethod = { ("toString" == name && "()Ljava/lang/String;" == descriptor) }

        !(isEqualsMethod() || isToStringMethod() || isHashCodeMethod())
    }

    /**
     * Static methods and Object-methods will be ignored
     *
     * @param methods original methods
     * @return filtered methods
     */
    fun filter(methods: Array<Method>) =
        methods.filter { !Modifier.isStatic(it.modifiers) }.filter { notObjectMethod(it) }.toList()
}
